kernel/kernel-api/src/main/java/io/delta/kernel/internal/TransactionImpl.java

```diff
@@ -145,18 +145,92 @@ public Optional<SetTransaction> getSetTxnOpt() {
     return setTxnOpt;
   }
 
+  public void addDomainMetadataInternal(String domain, String config) {
+    checkArgument(
+        !domainMetadatasRemoved.contains(domain),
+        "Cannot add a domain that is removed in this transaction");
+    checkState(!closed, "Cannot add a domain metadata after the transaction has completed");
+    // we override any existing value
+    domainMetadatasAdded.put(domain, new DomainMetadata(domain, config, false /* removed */));
+  }
+
+  @Override
+  public void addDomainMetadata(String domain, String config) {
+    checkArgument(
+        DomainMetadata.isUserControlledDomain(domain),
+        "Setting a system-controlled domain is not allowed: " + domain);
+    checkState(
+        TableFeatures.isDomainMetadataSupported(protocol),
+        "Unable to add domain metadata when the domain metadata table feature is disabled");
+    addDomainMetadataInternal(domain, config);
+  }
+
+  public void removeDomainMetadataInternal(String domain) {
+    checkArgument(
+        !domainMetadatasAdded.containsKey(domain),
+        "Cannot remove a domain that is added in this transaction");
+    checkState(!closed, "Cannot remove a domain after the transaction has completed");
+    domainMetadatasRemoved.add(domain);
+  }
+
+  @Override
+  public void removeDomainMetadata(String domain) {
+    checkArgument(
+        DomainMetadata.isUserControlledDomain(domain),
+        "Removing a system-controlled domain is not allowed: " + domain);
+    checkState(
+        TableFeatures.isDomainMetadataSupported(protocol),
+        "Unable to add domain metadata when the domain metadata table feature is disabled");
+    removeDomainMetadataInternal(domain);
+  }
+
   /**
-   * Internal API to add domain metadata actions for this transaction. Visible for testing.
+   * Returns a list of the domain metadatas to commit. This consists of the domain metadatas added
+   * in the transaction using {@link Transaction#addDomainMetadata(String, String)} and the
+   * tombstones for the domain metadatas removed in the transaction using {@link
+   * Transaction#removeDomainMetadata(String)}. The result is stored in {@code domainMetadatas}.
```

Comment:
Why do we store this in `domainMetadatas`? So we don't have to recall this for retries?

Reply:
I think maybe this method doesn't have to do this? the caller can set the variable

Reply:
and retry doesn't need to recall this, it can just use the variable since it must have already been resolved on first try

Reply:
ConflictChecker tries to get the domain metadatas using `transaction.getDomainMetadatas()`.

This was my attempt at minimizing the impact on the rest of the code. lmk if you have better alternatives though! :)

Reply:
Yeah I think this is okay; it's just a strange pattern to begin with since the dependent variables are mutating up to a certain point. I think we could put this inside of a lazy def for `domainMetadatas` but tbh that's not that different anyways.